### PR TITLE
Check in shim typeforwards and remove the dependency on the underlying targeting packs

### LINKED ISSUE #65997: Make .NET Framework and .NET Standard facade changes trackable
[GenFacade](https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.GenFacades) is a tool that inspects the satisfied and unsatisfied public API surface safe area based on a contract assembly and a passed in assembly closure. GenFacades is used to construct the .NETFramework and the .NETStandard shim assemblies which live under src/libraries/shims/.

The generated source file isn't checked into the tree, it's placed into the project's intermediate output path. IMO it would help if we check these files into the tree so that any changes that impact shims - intentional or unintentional - are trackable. As an example, I just recently refactored how the shims are structured under src/libraries/shims and unintentionally removed a type forward from an assembly. I only noticed this regression by pure luck later.

Additionally, adding the generated type-forward files to the tree would make it possible to construct the shims during the "libs.sfx" subset and later then, when the closure to compute the type forwards is complete during the "libs.oob" subset, invoke GenFacades to make sure that the generated files are up-to-date.

@dotnet/area-infrastructure-libraries @ericstj

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/runtime/issues/65997

This also changes the folder structure of shims to make them fit into the existing libraries layout (projectname/ref, projectname/src). That makes sharing msbuild data and C# files easier and enables the use of default compile items.

The shim projects now also define their dependencies explicitly instead of just receiving all references.

### CODE REVIEW

**Review by ericstj — COMMENTED** [review]
I'm supportive of checking in the forwards, but I think we need to think about the workflow of this and expectations for how developers interact with these sources.  Consider adding a readme describing the interaction.

**Review by carlossanlop — COMMENTED** [review]
I mostly have questions to learn about this.

Did the GenFacade tool generate both the cs files and csproj files under shims?

File: src/libraries/shims/Directory.Build.props [hunk 1]
```diff
@@ -19,44 +18,20 @@
...
-
-  <ItemGroup Condition="'$(MSBuildProjectName)' != 'netstandard'">
-    <!-- .NET Framework shims compile against OOBs and the netstandard shim. -->
-    <ProjectReference Include="$(MSBuildThisFileDirectory)src\netstandard.csproj" />
-    <ProjectReference Include="$(MSBuildThisFileDirectory)..\oob-src.proj" Condition="'$(RefOnly)' != 'true'" />
-    <ProjectReference Include="$(MSBuildThisFileDirectory)..\oob-ref.proj" Condition="'$(RefOnly)' == 'true'" />
-  </ItemGroup>
-
-  <ItemGroup>
-    <!-- System.Data.SqlClient is not live built anymore, reference it manually to avoid tfm nuget selection fallback errors
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/shims/Directory.Build.props:52]: This reference was moved directly into shims/System.Data/Directory.Build.props

File: src/libraries/shims/System/Directory.Build.props [hunk 2]
```diff
@@ -1,6 +1,10 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project>
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/shims/System/Directory.Build.props:1]: That's just git that gets confused with the file changes...

File: src/libraries/shims/Directory.Build.props [hunk 3]
```diff
@@ -8,6 +8,8 @@
     <!-- By default make shims compile against reference assemblies. -->
     <CompileUsingReferenceAssemblies>true</CompileUsingReferenceAssemblies>
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
+    <GenFacadesOutputSourcePath>$(MSBuildProjectDirectory)\$(MSBuildProjectName).g.cs</GenFacadesOutputSourcePath>
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/shims/Directory.Build.props]: Sorry Eric, I should have changed the PR's state to WIP. You reviewed it a little bit too early. It should now be ready.

File: src/libraries/shims/Directory.Build.props [hunk 4]
```diff
@@ -4,59 +4,39 @@
...
-    <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
-    <GenFacadesIgnoreMissingTypes>true</GenFacadesIgnoreMissingTypes>
-    <!-- ensure the desktop compat shims reference the lowest possible version of dependencies
-         since those do not all ship as part of the framework and we don't want to force apps
-         to reference the latest packages.
-         netstandard.dll doesn't need to do this since it has no dangling dependencies -->
-    <GenFacadesForceZeroVersionSeeds Condition="'$(MSBuildProjectName)' != 'netstandard'">true</GenFacadesForceZeroVersionSeeds>
+    <!-- Ensure the .NET Framework shims reference the lowest possible version of dependencies since
+         those do not all ship as part of the framework and we don't want to force apps to reference the
+         latest packages. netstandard.dll doesn't need to do this since it has no dangling dependencies. -->
```
Comment (carlossanlop) [inline src/libraries/shims/Directory.Build.props:12]: Why do we not want the latest package to he referenced?

Comment (ViktorHofer) [reply src/libraries/shims/Directory.Build.props:12]: This is already the same in main. Or are you asking about why we are doing this generally? IMO the comment captures that nicely. We make sure that shims don't depend on the very latest packages and instead work with any version. When you look at any of the shims in ilspy, i.e. mscorlib you will see that all dependency versions in its metadata are 0.0.0.0.

File: src/libraries/shims/System.Xml.Serialization/src/System.Xml.Serialization.csproj [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <StrongNameKeyId>ECMA</StrongNameKeyId>
```
Comment (carlossanlop) [inline src/libraries/shims/System.Xml.Serialization/src/System.Xml.Serialization.csproj:5]: I see we use Microsoft, MicrosoftShared or ECMA values. What is the purpose of using difference values? How do you decide which one to use?

**Review by carlossanlop — APPROVED** [review]
Assuming the CI does not contain related failures, LGTM.

File: src/libraries/shims/System.Xml.Serialization/src/System.Xml.Serialization.csproj (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/shims/System.Xml.Serialization/src/System.Xml.Serialization.csproj:5]: These assemblies exist in the .NET Framework runtime on Windows machines and need to have the same identity. The strong name key is part of the assembly's identity, same as the version. We just pick whatever strong name key was chosen for the assembly in the .NET Framework runtime. Note that this also already existed in main, I didn't change any of these values, just moved them. Git diff unfortunately doesn't help here.

**Pull request merged**